### PR TITLE
Add no-op JVM target

### PR DESCRIPTION
**Pull request by konpach**
Relates Issue: #380

Context: Until library gets a proper implementation for jvm targets we want to allow the usage of the library in supported targets along with jvm. Currently is not possible to compile shared if jvm target exists (as there is no corresponding implementation in kable).
Solution: Provide an empty implementation that throws runtime exceptions for kable for jvm target.

### CODE REVIEW

**Comment by konpach** [comment]
Build failed due to 
```
Expected folder with API declarations '/Users/runner/work/kable/kable/core/api/android' does not exist.
  Please ensure that ':apiDump' was executed in order to get API dump to compare the build against
```
I thought that these are generated during the build, will run locally and add.

**Comment by twyatt** [comment]
This may have changed/improved, but I thought historically having a JVM _and_ Android target in the same module would confuse Gradle on the consumer side?

In other words, would an Android app project be able to consume Kable if it had both an Android target and a (no-op) JVM target? How would it know to use the Android when both are available?

If you have a chance (if not, then I'll try to find some time) to test these changes in the [SensorTag sample app](https://github.com/JuulLabs/sensortag) to ensure it doesn't break Android app consumers.

**Comment by konpach** [comment]
> This may have changed/improved, but I thought historically having a JVM _and_ Android target in the same module would confuse Gradle on the consumer side?
> 
> In other words, would an Android app project be able to consume Kable if it had both an Android target and a (no-op) JVM target? How would it know to use the Android when both are available?
> 
TBH I haven't experienced this before. I thought it's handled by gradle. But good point

> If you have a chance (if not, then I'll try to find some time) to test these changes in the [SensorTag sample app](https://github.com/JuulLabs/sensortag) to ensure it doesn't break Android app consumers.
> 
> If it does break Android consumers, then I believe we'll have to introduce a new `-android` suffixed artifact for Android consumers. 🤔

I tried it with SensorTag and it works as expected. Also, I tried it with a project that has iOS, android and desktop targets and it compiles and runs as expected for all of the platforms. After you comment I also did a sanity check with my android app that scanning and connecting to a BLE device works as before (without the no-op jvm impl).

@twyatt

**Comment by twyatt** [comment]
Awesome! Thanks for testing/validating!

**Review by cedrickcooke — APPROVED** [review]
Looks good to me.

Weighing in on the build issues re. having both `jvm` and `android` targets, I think I remember it being a problem specifically when there was a diamond hierarchy. If `jvm-library` depends on `kable, `android-library` depends on `kable`, and `your-android-app` depends on both `jvm-library` and `android-library`, problems show up.

Supporting JVM doesn't automatically break anything so I'm cool with this going in, but it does leave some room to shoot yourself in the foot.

File: core/src/jvmMain/kotlin/com/juul/kable/logs/SystemLogEngine.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package com.juul.kable.logs
+
+import com.juul.kable.jvmNotImplementedException
+
+public actual object SystemLogEngine : LogEngine {
+
+    override fun verbose(throwable: Throwable?, tag: String, message: String) {
+        debug(throwable, tag, message)
+    }
```
Comment (twyatt) [inline core/src/jvmMain/kotlin/com/juul/kable/logs/SystemLogEngine.kt:9]: Particular reason for this function to call to `debug` but the others to call `jvmNotImplementedException()`?

Comment (konpach) [reply core/src/jvmMain/kotlin/com/juul/kable/logs/SystemLogEngine.kt:9]: No, this slipped away. Fixing it.

File: core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.juul.kable
+
+public fun jvmNotImplementedException(): Nothing = throw NotImplementedError(
```
[the remark this answers is not in the document]
Comment (konpach) [reply core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt:3]: No, if it's internal compiler gives error about the visibility.

File: core/src/androidMain/kotlin/Profile.kt [hunk 3]
```diff
@@ -38,7 +38,7 @@ public actual data class DiscoveredCharacteristic internal constructor(
     override val serviceUuid: Uuid get() = characteristic.service.uuid
     override val characteristicUuid: Uuid get() = characteristic.uuid
     val instanceId: Int get() = characteristic.instanceId
-    actual val properties: Properties get() = Properties(characteristic.properties)
+    public actual val properties: Properties get() = Properties(characteristic.properties)
```
[the remark this answers is not in the document]
Comment (konpach) [reply core/src/androidMain/kotlin/Profile.kt:41]: No, I just tried to align it with the visibility of this field in the implementations of other targets.

**Comment by konpach** [comment]
@twyatt anything more to look into?

File: core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (twyatt) [reply core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt:3]: I presume the error is related to other modules not being about to access it?

In that case, can we make it `internal` and duplicate the function in any module that uses it?

That way we aren't exposing a new function on the public API.

**Comment by twyatt** [comment]
Apologies..

Looks good aside from one last comment I left re: removing the expectation function from the public API.

File: core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (konpach) [reply core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt]: Good point. Actually, I miss understood the error. It seemed relevant but actually it was hiding a missing impl in the exceptions module! Made this internal and also added the  implementation for the jvm target for the IOException.

[the remark this answers is not in the document]
Comment (konpach) [reply core/src/jvmMain/kotlin/com/juul/kable/Exceptions.kt]: Saw your revert, makes sense, thank you. Don't know why building locally failed but passing here in GHA. Will try to understand.

**Review by twyatt — APPROVED** [review]
Looks great!
I'll do some final testing this weekend and get it merged/released.
Thanks @konpach!

**Pull request merged**